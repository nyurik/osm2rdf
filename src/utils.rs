use chrono::{TimeZone, Utc};
use json::JsonValue;
use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
use regex::Regex;

pub(crate) trait StringExt {
    fn push_str_value(&mut self, predicate: &'static str, value: &str);
    fn push_quoted_str(&mut self, value: &str);
    fn push_char_value(&mut self, predicate: &'static str, value: char);
    fn push_bool_value(&mut self, predicate: &'static str, value: bool);
    fn push_int_value(&mut self, predicate: &'static str, value: i64);
    fn push_date_value(&mut self, predicate: &'static str, value: i64);
    fn push_point(&mut self, predicate: &'static str, latitude: f64, longitude: f64);
    fn push_wiki_url(&mut self, lang: &str, site: &str, title: &str);
    fn push_tag(&mut self, key: &str, value: &str, utils: &Consts);
    fn push_metadata(&mut self, version: i32, user: &str, milli_timestamp: i64, changeset: i64);
}

impl StringExt for String {
    fn push_str_value(&mut self, predicate: &'static str, value: &str) {
        self.push_str(predicate);
        self.push(' ');
        self.push_quoted_str(value);
    }

    fn push_quoted_str(&mut self, value: &str) {
        // TODO: optimize?
        self.push_str(&JsonValue::from(value).dump());
        self.push_str(";\n");
    }

    fn push_char_value(&mut self, predicate: &'static str, value: char) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push(value);
        self.push_str("\";\n");
    }

    fn push_bool_value(&mut self, predicate: &'static str, value: bool) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(if value { "true" } else { "false" });
        self.push_str("\"^^xsd:boolean;\n");
    }

    fn push_int_value(&mut self, predicate: &'static str, value: i64) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(&value.to_string());
        self.push_str("\"^^xsd:integer;\n");
    }

    fn push_date_value(&mut self, predicate: &'static str, milli_timestamp: i64) {
        // "{0:%Y-%m-%dT%H:%M:%S}Z"^^xsd:dateTime
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(&format_ts(milli_timestamp));
        self.push_str("\"^^xsd:dateTime;\n");
    }

    fn push_point(&mut self, predicate: &'static str, latitude: f64, longitude: f64) {
        self.push_str(predicate);
        self.push_str(" \"Point(");
        self.push_str(&longitude.to_string());
        self.push(' ');
        self.push_str(&latitude.to_string());
        self.push_str(")\"^^geo:wktLiteral");
        self.push_str(";\n");
    }

    fn push_wiki_url(&mut self, lang: &str, site: &str, title: &str) {
        self.push_str("<https://");
        self.push_str(lang);
        self.push_str(site);
        self.extend(utf8_percent_encode(
            title.replace(' ', "_").as_str(),
            PERCENT_ENC_SET,
        ));
        self.push('>');
        self.push_str(";\n");
    }

    fn push_tag(&mut self, key: &str, value: &str, consts: &Consts) {
        if !consts.re_simple_local_name.is_match(key) {
            // Record any unusual tag name in a "osmm:badkey" statement
            self.push_str_value("osmm:badkey", value);
            return;
        }

        self.push_str("osmt:");
        self.push_str(key);
        self.push(' ');
        let mut parsed = false;
        if key.contains("wikidata") {
            if consts.re_wikidata_value.is_match(value) {
                self.push_str("wd:");
                self.push_str(value);
                parsed = true;
            } else if consts.re_wikidata_multi_value.is_match(value) {
                for v in value.split(';') {
                    self.push_str("wd:");
                    self.push_str(v);
                    self.push(',');
                }
                self.pop(); // remove trailing ","
                parsed = true;
            }
        } else if key.contains("wikipedia") {
            if let Some(v) = consts.re_wikipedia_value.captures(value) {
                self.push_wiki_url(
                    v.get(1).unwrap().as_str(),
                    ".wikipedia.org/wiki/",
                    v.get(2).unwrap().as_str(),
                );
                parsed = true;
            }
        }
        if !parsed {
            self.push_quoted_str(value);
        }
    }

    fn push_metadata(&mut self, version: i32, user: &str, milli_timestamp: i64, changeset: i64) {
        self.push_int_value("osmm:version", version as i64);
        self.push_str_value("osmm:user", user);
        self.push_date_value("osmm:timestamp", milli_timestamp);
        self.push_int_value("osmm:changeset", changeset);
        self.pop();
        self.pop();
        self.push_str(".\n");
    }
}

pub fn format_ts(milli_timestamp: i64) -> String {
    format!(
        "{:?}",
        Utc.timestamp(milli_timestamp / 1000, (milli_timestamp % 1000) as u32)
    )
}

pub const PERCENT_ENC_SET: &AsciiSet = &CONTROLS
    .add(b';')
    .add(b'@')
    .add(b'$')
    .add(b'!')
    .add(b'*')
    .add(b'(')
    .add(b')')
    .add(b',')
    .add(b'/')
    .add(b'~')
    .add(b':')
    // The "#" is also safe - used for anchoring
    .add(b'#');

#[derive(Clone, Debug)]
pub struct Consts {
    pub re_simple_local_name: Regex,
    pub re_wikidata_key: Regex,
    pub re_wikidata_value: Regex,
    pub re_wikidata_multi_value: Regex,
    pub re_wikipedia_value: Regex,
}

impl Consts {
    pub fn new() -> Self {
        Consts {
            /// Total length of the maximum "valid" local name is 60 (58 + first + last char)
            /// Local name may contain letters, numbers anywhere, and -:_ symbols anywhere except first and last position
            re_simple_local_name: Regex::new(r"^[0-9a-zA-Z_]([-:0-9a-zA-Z_]{0,58}[0-9a-zA-Z_])?$")
                .unwrap(),

            re_wikidata_key: Regex::new(r"(.:)?wikidata$").unwrap(),
            re_wikidata_value: Regex::new(r"^Q[1-9][0-9]{0,18}$").unwrap(),
            re_wikidata_multi_value: Regex::new(r"^Q[1-9][0-9]{0,18}(;Q[1-9][0-9]{0,18})+$")
                .unwrap(),
            re_wikipedia_value: Regex::new(r"^([-a-z]+):(.+)$").unwrap(),
        }
    }
}

#[derive(Clone, Default, Debug)]
pub struct Stats {
    pub added_nodes: u64,
    pub added_rels: u64,
    pub added_ways: u64,
    pub skipped_nodes: u64,
    pub deleted_nodes: u64,
    pub deleted_rels: u64,
    pub deleted_ways: u64,
    pub blocks: u64,
}

impl Stats {
    pub(crate) fn combine(&mut self, other: Stats) {
        self.added_nodes += other.added_nodes;
        self.added_rels += other.added_rels;
        self.added_ways += other.added_ways;
        self.skipped_nodes += other.skipped_nodes;
        self.deleted_nodes += other.deleted_nodes;
        self.deleted_rels += other.deleted_rels;
        self.deleted_ways += other.deleted_ways;
        self.blocks += 1;
    }
}

#[derive(Debug)]
pub enum Statement {
    Skip,
    Delete {
        elem: Element,
        id: i64,
    },
    Create {
        elem: Element,
        id: i64,
        ts: i64,
        val: String,
    },
}

#[derive(Debug)]
pub enum Element {
    Node,
    Way,
    Relation,
}
