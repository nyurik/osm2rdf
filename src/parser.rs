use std::fs::File;
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicI64, AtomicU32, Ordering};
use std::sync::mpsc::{channel, Receiver, Sender};
use std::sync::Mutex;
use std::thread::{Builder, JoinHandle};

use bytesize::ByteSize;
use flate2::write::GzEncoder;
use flate2::Compression;
use geos::{CoordSeq, Geom, Geometry};
use lazy_static::lazy_static;
use osmnodecache::{Cache, CacheStore, DenseFileCache, DenseFileCacheOpts, HashMapCache};
use osmpbf::{BlobDecode, BlobReader, DenseNode, Info, Node, PrimitiveBlock, Relation, Way};
use path_absolutize::Absolutize;
use percent_encoding::utf8_percent_encode;
use rayon::iter::{ParallelBridge, ParallelIterator};
use regex::Regex;

use crate::str_builder::{
    StringBuf, XsdBoolean, XsdElement, XsdIter, XsdPoint, XsdRaw, XsdRelMember, XsdStr,
    XsdWikipedia,
};
use crate::utils::{to_utc, Element, Statement, Stats, PERCENT_ENC_SET};
use crate::{Args, Command};

lazy_static! {
    /// Total length of the maximum "valid" local name is 60 (58 + first + last char)
    /// Local name may contain letters, numbers anywhere, and -:_ symbols anywhere except first and last position
    pub static ref RE_SIMPLE_LOCAL_NAME: Regex = Regex::new(r"^[0-9a-zA-Z_]([-:0-9a-zA-Z_]{0,58}[0-9a-zA-Z_])?$").unwrap();
    pub static ref RE_WIKIDATA_KEY: Regex = Regex::new(r"(.:)?wikidata$").unwrap();
    pub static ref RE_WIKIDATA_VALUE: Regex = Regex::new(r"^Q[1-9][0-9]{0,18}$").unwrap();
    pub static ref RE_WIKIDATA_MULTI_VALUE: Regex = Regex::new(r"^Q[1-9][0-9]{0,18}(\s*;\s*Q[1-9][0-9]{0,18})+$").unwrap();
    pub static ref RE_WIKIPEDIA_VALUE: Regex = Regex::new(r"^([-a-z]+):(.+)$").unwrap();
}

pub struct Parser<'a> {
    parent_stats: &'a Mutex<Stats>,
    stats: Stats,
    cache: Box<dyn Cache + 'a>,
}

impl<'a> Drop for Parser<'a> {
    fn drop(&mut self) {
        let stats = std::mem::take(&mut self.stats);
        self.parent_stats.lock().unwrap().combine(stats);
    }
}

impl<'a> Parser<'a> {
    pub fn new(parent_stats: &'a Mutex<Stats>, cache: Box<dyn 'a + Cache>) -> Parser<'a> {
        Parser {
            parent_stats,
            stats: Stats::default(),
            cache,
        }
    }

    fn on_node(&mut self, node: &Node) -> Statement {
        let info = node.info();
        let mut statement = self.process_node(
            info.deleted(),
            node.id(),
            node.tags(),
            node.lat(),
            node.lon(),
        );
        if let Statement::Create {
            ref mut val,
            ref mut ts,
            ..
        } = statement
        {
            *ts = Self::push_info(val, &info);
        }
        statement
    }

    fn on_dense_node(&mut self, node: &DenseNode) -> Statement {
        let info = node.info().unwrap();
        let mut statement = self.process_node(
            info.deleted(),
            node.id(),
            node.tags(),
            node.lat(),
            node.lon(),
        );
        if let Statement::Create {
            val: ref mut value, ..
        } = statement
        {
            value.push_metadata(
                info.version(),
                info.user().unwrap(),
                info.milli_timestamp(),
                info.changeset(),
            );
        }
        statement
    }

    fn process_node<'t, TTags: Iterator<Item = (&'t str, &'t str)> + ExactSizeIterator>(
        &mut self,
        is_deleted: bool,
        id: i64,
        tags: TTags,
        lat: f64,
        lon: f64,
    ) -> Statement {
        if is_deleted {
            self.stats.deleted_nodes += 1;
            Statement::Delete {
                elem: Element::Node,
                id,
            }
        } else {
            self.cache.set_lat_lon(id as usize, lat, lon);
            let mut value = StringBuf::new(100000);
            self.push_all_tags(&mut value, tags);
            if value.is_empty() {
                self.stats.skipped_nodes += 1;
                Statement::Skip
            } else {
                value.push_value("osmm:loc", XsdPoint { lat, lon });
                value.push_value("osmm:type", XsdElement(Element::Node));
                self.stats.added_nodes += 1;
                Statement::Create {
                    elem: Element::Node,
                    id,
                    val: value,
                    ts: 0,
                }
            }
        }
    }

    fn on_way(&mut self, way: &Way) -> Statement {
        let info = way.info();
        if info.deleted() {
            self.stats.deleted_ways += 1;
            return Statement::Delete {
                elem: Element::Way,
                id: way.id(),
            };
        }
        let mut value = StringBuf::new(100000);
        self.push_all_tags(&mut value, way.tags());
        value.push_value("osmm:type", XsdElement(Element::Way));
        let ts = Self::push_info(&mut value, &info);
        if let Err(err) = self.parse_way_geometry(&mut value, way) {
            value.push_value("osmm:loc:error", XsdStr(&err.to_string()));
        }

        self.stats.added_ways += 1;
        Statement::Create {
            elem: Element::Way,
            id: way.id(),
            val: value,
            ts,
        }
    }

    fn on_relation(&mut self, rel: &Relation) -> Statement {
        let info = rel.info();
        if info.deleted() {
            self.stats.deleted_rels += 1;
            return Statement::Delete {
                elem: Element::Way,
                id: rel.id(),
            };
        }
        let mut value = StringBuf::new(100000);
        self.push_all_tags(&mut value, rel.tags());
        value.push_value("osmm:type", XsdElement(Element::Relation));

        let ts = Self::push_info(&mut value, &info);
        for mbr in rel.members() {
            // Produce two statements - one to find all members of a relation,
            // and another to find the role of that relation
            //     osmrel:123  osmm:has    osmway:456
            //     osmrel:123  osmway:456  "inner"
            value.push_value("osmm:has", XsdRelMember(&mbr));
            value.push_value(XsdRelMember(&mbr), XsdStr(mbr.role().unwrap()));
        }

        self.stats.added_rels += 1;
        Statement::Create {
            elem: Element::Relation,
            id: rel.id(),
            val: value,
            ts,
        }
    }

    fn push_info(value: &mut StringBuf, info: &Info) -> i64 {
        let ts = info.milli_timestamp().unwrap();
        value.push_metadata(
            info.version().unwrap(),
            info.user().unwrap().unwrap(),
            ts,
            info.changeset().unwrap(),
        );
        ts
    }

    fn parse_way_geometry(&self, value: &mut StringBuf, way: &Way) -> anyhow::Result<()> {
        let refs: Vec<[f64; 2]> = way
            .refs()
            .map(|id| {
                let (lat, lng) = self.cache.get_lat_lon(id as usize);
                [lat, lng]
            })
            .collect();

        let geometry = Geometry::create_line_string(CoordSeq::new_from_vec(&refs)?)?;
        let value1 = geometry.is_closed()?;
        value.push_value("osmm:isClosed", XsdBoolean(value1));
        let g = geometry.point_on_surface()?;
        let lat = g.get_y().unwrap();
        let lon = g.get_x().unwrap();
        value.push_value("osmm:loc", XsdPoint { lat, lon });

        Ok(())
    }

    fn push_all_tags<'t, TTags: Iterator<Item = (&'t str, &'t str)> + ExactSizeIterator>(
        &mut self,
        value: &mut StringBuf,
        tags: TTags,
    ) {
        for (key, val) in tags {
            if key == "created_by" {
                continue;
            }
            if !RE_SIMPLE_LOCAL_NAME.is_match(key) {
                // Record any unusual tag name in a "osmm:badkey" statement
                value.push_value("osmm:badkey", XsdStr(val));
                continue;
            }

            let prop = XsdRaw("osmt", key);
            if key.contains("wikidata") {
                if RE_WIKIDATA_VALUE.is_match(val) {
                    value.push_value(prop, XsdRaw("wd", val));
                    continue;
                } else if RE_WIKIDATA_MULTI_VALUE.is_match(val) {
                    value.push_value(
                        prop,
                        XsdIter(|| val.split(';').map(|v| XsdRaw("wd", v.trim()))),
                    );
                    continue;
                }
            } else if key.contains("wikipedia") {
                if let Some(v) = RE_WIKIPEDIA_VALUE.captures(val) {
                    let lang = v.get(1).unwrap().as_str();
                    let title = v.get(2).unwrap().as_str();
                    let title = title.replace(' ', "_");
                    let title = &utf8_percent_encode(&title, PERCENT_ENC_SET);
                    value.push_value(prop, XsdWikipedia { lang, title });
                    continue;
                }
            }
            value.push_value(prop, XsdStr(val));
        }
    }
}

fn create_flat_cache(filename: PathBuf) -> anyhow::Result<DenseFileCache> {
    Ok(DenseFileCacheOpts::new(filename)
        .page_size(10 * 1024 * 1024 * 1024)
        .on_size_change(Some(|old_size, new_size| {
            println!(
                "Growing cache {} âž¡ {}",
                ByteSize(old_size as u64),
                ByteSize(new_size as u64)
            )
        }))
        .open()?)
}

fn start_writer_thread(
    output_dir: &Path,
    max_file_size: usize,
    receiver: Receiver<Statement>,
) -> JoinHandle<()> {
    let output_dir = output_dir.to_path_buf();
    let file_index = AtomicU32::new(0);
    let oldest_ts = AtomicI64::new(0);

    Builder::new()
        .name("gz_writer".into())
        .spawn(move || {
            let mut encoder = None;
            let mut size = 0_usize;
            while let Ok(v) = receiver.recv() {
                if let Statement::Create { elem, id, val, ts } = v {
                    oldest_ts.fetch_max(ts, Ordering::Relaxed);

                    let enc = encoder.get_or_insert_with(|| new_gz_file(&output_dir, &file_index));
                    writeln!(enc, "{elem}:{id}\n{val}").unwrap();

                    size += val.len();
                    if size > max_file_size {
                        encoder.take().unwrap().finish().unwrap();
                        size = 0;
                    }
                }
            }

            let mut enc = new_gz_file(&output_dir, &file_index);
            let ts = to_utc(oldest_ts.load(Ordering::SeqCst));
            writeln!(enc, "osmroot: schema:dateModified {ts}.").unwrap();
        })
        .unwrap()
}

fn new_gz_file(output_dir: &Path, file_index: &AtomicU32) -> GzEncoder<File> {
    let index = file_index.fetch_add(1, Ordering::Relaxed);
    let file = output_dir.join(format!("osm-{index:06}.ttl.gz"));
    println!("Creating {:?}", file.absolutize().unwrap());
    GzEncoder::new(File::create(file).unwrap(), Compression::default())
}

pub fn parse(opt: Args) -> anyhow::Result<()> {
    match opt.cmd {
        Command::Parse {
            workers,
            input_file,
            output_dir,
            max_file_size,
        } => {
            if let Some(v) = workers {
                rayon::ThreadPoolBuilder::new()
                    .thread_name(|i| format!("parser #{i}"))
                    .num_threads(v)
                    .build_global()
                    .unwrap();
            }
            let (sender, receiver) = channel();
            let writer_thread =
                start_writer_thread(&output_dir, max_file_size * 1024 * 1024, receiver);

            let reader = BlobReader::from_path(input_file)?;
            if let Some(filename) = &opt.planet_cache {
                let cache = create_flat_cache(filename.clone())?;
                parse_with_cache(cache, sender, reader);
            } else {
                let filename = &opt.small_cache.unwrap();
                let cache = if filename.exists() {
                    HashMapCache::from_bin(filename)?
                } else {
                    HashMapCache::new()
                };
                parse_with_cache(cache.clone(), sender, reader);
                cache.save_as_bin(filename)?;
            }

            writer_thread.join().unwrap();
            Ok(())
        } // _ => panic!("Expecting Parse")
    }
}

pub fn parse_with_cache<R: Read + Send, C: CacheStore + Clone + Send>(
    cache: C,
    sender: Sender<Statement>,
    reader: BlobReader<R>,
) {
    let stats = Mutex::new(Stats::default());
    reader
        .par_bridge()
        .for_each_with((cache, sender), |(dfc, writer), blob| {
            if let BlobDecode::OsmData(block) = blob.unwrap().decode().unwrap() {
                let mut parser = Parser::new(&stats, dfc.get_accessor());
                parse_block(block, &mut parser, |s| writer.send(s).unwrap());
            };
        });
    println!("{:#?}", stats.lock().unwrap());
}

pub fn parse_block(block: PrimitiveBlock, parser: &mut Parser, mut writer: impl FnMut(Statement)) {
    for group in block.groups() {
        // FIXME: possible concurrency bug: a non-node element may need coords of a node that hasn't been processed yet
        for node in group.nodes() {
            writer(parser.on_node(&node));
        }
        for node in group.dense_nodes() {
            writer(parser.on_dense_node(&node));
        }
        for way in group.ways() {
            writer(parser.on_way(&way));
        }
        for rel in group.relations() {
            writer(parser.on_relation(&rel));
        }
    }
}
