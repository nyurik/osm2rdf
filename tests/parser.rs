use insta::glob;
use osm2rdf::parser::parse_block;
use osm2rdf::parser::Parser;
use osm2rdf::utils::Stats;
use osmnodecache::{CacheStore, HashMapCache};
use osmpbf::{BlobDecode, BlobReader};
use std::sync::Mutex;

#[test]
fn decode_osm_pbf_files() {
    glob!("../tests/fixtures", "**/*.osm.pbf", |file| {
        let reader = BlobReader::from_path(file).unwrap();
        let cache = HashMapCache::new();
        let stats = Mutex::new(Stats::default());
        let mut parser = Parser::new(&stats, cache.get_accessor());

        let mut result = Vec::new();
        for blob in reader {
            if let BlobDecode::OsmData(block) = blob.unwrap().decode().unwrap() {
                parse_block(block, &mut parser, |s| result.push(s));
            };
        }
        insta::assert_debug_snapshot!(result);
    });
}
