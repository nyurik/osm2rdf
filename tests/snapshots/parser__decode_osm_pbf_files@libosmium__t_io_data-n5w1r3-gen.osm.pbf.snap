---
source: tests/parser.rs
expression: result
input_file: tests/fixtures/libosmium/t_io_data-n5w1r3-gen.osm.pbf
---
[
    Skip,
    Skip,
    Skip,
    Skip,
    Skip,
    Create {
        elem: Way,
        id: 20,
        ts: 1388534400000,
        val: indoc! {r#"
            osmt:highway "primary";
            osmm:type "w";
            osmm:version "1"^^xsd:integer;
            osmm:user "test";
            osmm:timestamp "2014-01-01 00:00:00 UTC"^^xsd:dateTime;
            osmm:changeset "1"^^xsd:integer.
            osmm:isClosed "false"^^xsd:boolean;
            osmm:loc "Point(0.9999999827705324 1.0999999852385371)"^^geo:wktLiteral;
        "}#",
    },
    Create {
        elem: Relation,
        id: 30,
        ts: 1388534400000,
        val: indoc! {r#"
            osmt:type "test";
            osmm:type "r";
            osmm:version "1"^^xsd:integer;
            osmm:user "test";
            osmm:timestamp "2014-01-01 00:00:00 UTC"^^xsd:dateTime;
            osmm:changeset "1"^^xsd:integer.
            osmm:has osmnode:10;
            osmnode:10 "none";
        "}#",
    },
    Create {
        elem: Relation,
        id: 31,
        ts: 1388534400000,
        val: indoc! {r#"
            osmt:type "restriction";
            osmm:type "r";
            osmm:version "1"^^xsd:integer;
            osmm:user "test";
            osmm:timestamp "2014-01-01 00:00:00 UTC"^^xsd:dateTime;
            osmm:changeset "1"^^xsd:integer.
            osmm:has osmway:20;
            osmway:20 "from";
            osmm:has osmnode:11;
            osmnode:11 "via";
            osmm:has osmway:22;
            osmway:22 "to";
        "}#",
    },
    Create {
        elem: Relation,
        id: 32,
        ts: 1388534400000,
        val: indoc! {r#"
            osmt:test "relinrel";
            osmm:type "r";
            osmm:version "1"^^xsd:integer;
            osmm:user "test";
            osmm:timestamp "2014-01-01 00:00:00 UTC"^^xsd:dateTime;
            osmm:changeset "1"^^xsd:integer.
            osmm:has osmrel:30;
            osmrel:30 "none";
        "}#",
    },
]
